### PR TITLE
Event fixes and improvements (Console)

### LINKED ISSUE #2887: Data views slow down and freeze after receiving many events
#### Summary
The data views of the console freeze after receiving a lot of events.

#### Steps to Reproduce
1. Go to the data view for an entity that receives a lot of events
2. Wait for some time while observing incoming events (waiting time is relative to the event frequency)
3. Observe the app becoming slow
4. Observe the app eventually freezing completely

#### What do you see now?
Freezing data views

#### What do you want to see instead?
Data views being stable regardless of the amount of events received

#### Environment
v3.5.8 and older

#### How do you propose to implement this?
The problem is likely caused by the redux store being filled considerably when the console is subscribed to an event source. Especially since the packet broker update, it will receive a lot of data per event which easily sucks up memory. My ideas to fix this are:
- Introduce a maximum amount of messages that are stored (rest being truncated)
- Discard event data for old events
- Experiment with storing event data in the browsers `localStorage` instead of in the redux store. We could then only reference the event data in the redux store. This is assumption-based buy maybe the `localStorage` can handle large amounts of data better.

@bafonins what do you think?

#### How do you propose to test this?
Manual testing.

#### Can you do this yourself and submit a Pull Request?
Yes.

### LINKED ISSUE #2980: Show all event info in expanded event details
#### Summary
The expanded details section of the console currently only shows the `data` prop of the event. However, there is other relevant data in the root of the event object that should be shown too.

#### Why do we need this?
The user should be able to inspect all available data of an event if he/she wishes to

#### What is already there? What do you see now?
Only the `data` prop of the event object being shown

#### What is missing? What do you want to see?
The whole event object being shown instead.

@bafonins @johanstokking has there been any discussion that led to this decision that I might have missed?

#### Environment
v3.8.x

#### How do you propose to implement this?
Just change the details component to show `event` instead of `event.data`

#### How do you propose to test this?
Trivial.

#### Can you do this yourself and submit a Pull Request?
Yes.

### LINKED ISSUE #3098: Event list does not account for multiple events with same name and timestamp
#### Summary
The event list component is prone to same key warnings, happening when occasionally multiple events with the same name and timestamps are generated.

#### Steps to Reproduce
1. Go to gateway events (it's likelier to happen here but this can happen also for applicaiton/end device events)
2. Wait for some events to arrive
3. Eventually, see a duplicate key warning in the logs

#### What do you see now?
```
react-dom.development.js?7f13:546 Warning: Encountered two children with the same key, `2020-08-18T15:25:49.148698549Z-gs.status.forward`.
```

#### What do you want to see instead?
Event keys being composed in a way that is sufficiently unique.

#### Environment
v3.9.0

#### How do you propose to implement this?
We need to find a way to derive a truly UID from the event content. If some events are actually identical then this will be impossible, but on the other hand raises the question of why such events exist in the first place

#### How do you propose to test this?
Manual.

#### Can you do this yourself and submit a Pull Request?
Yes. I'll fix this as part of my work on #2826

### PR DESCRIPTION
**Pull request by kschiffer**
#### Summary
This PR introduces fixes and improvements to event components and  views. It solves performance issues (#2887), [layout issues](https://github.com/TheThingsIndustries/lorawan-stack/issues/2260#issuecomment-682060901)), moves to show the complete event data on expand (#2980) and introduces "synthetic" status events (see #2884).

Closes #2887
Closes #2980
References #2260
References #2884

<details><summary>Screenshot</summary>

image
</details>

#### Changes
- Refactor event global and console components into a single event component
- Determine the display of the data preview column by the data type
- Add specific preview components for most common data message type
- Move to showing the event details on the side, detached from the event row
- Introduce `react-window` to display the list to only render the currently visible events
- Add utilities and definitions of synthetic events to display frontend defined messages in the event stream
- Add definitions and display logic for synthetic status events (will be used in upcoming PR)
- Add utility to `<ErrorMessage />` to display the root cause together with the top level message
- Reorganize hardcoded icons

#### Testing

I did quite some manual testing using the staging environment with live data as well as storybook. End-to-end testing for events is still not possible since we would need a workflow to simulate events inside test.

##### Regressions

This PR could cause regressions regarding:
- Event display
- Event stream error handling

#### Notes for Reviewers
The diff unfortunately became quite large, even after branching off some of the changes. Since I decided to refactor a lot of the code fundamentally, all things considered, I don't think any more splitting up will do us a favor.

- This PR already contains definitions for synthetic events that are not yet used. I decided against adding the store updates to improve the error resilience (#2884) in this PR to reduce the diff so I'll add that in a separate PR once this is merged
- I removed the color coding of events that we currently have. As can be seen in [#2260](https://github.com/TheThingsIndustries/lorawan-stack/issues/2260#issuecomment-682060901) this can cause false signals of caution for events that are actually harmless. This needs some better consideration which I will track in a new issue.
- I decided to use the message type (`event.data['@type']`) to determine the data preview since I believe this is clearer and easier to maintain than going from the event name.
  - The previews that I chose for particular message types are based on what we currently display but these need to be determined more clearly based on some feedback and discussions
- This PR unifies the fragmented, repetitive and as I think somewhat too abstracted structure that we have for event components and containers into just one `<Events />` component. I believe this improves the maintainability of this core functionality of the console.
- I have removed the error display for event errors that were caught by the error boundary as these errors are frontend-specific and not helpful for the user. Instead these errors now get logged to sentry.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/webui/console/components/events/index.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import Widget from './widget'
+import m from './messages'
+import { getEventId } from './utils'
+
+import style from './events.styl'
+
+const Events = React.memo(({ events, scoped, onClear, entityId }) => {
+  const [paused, setPause] = useState(false)
+  const [focus, setFocus] = useState({ eventId: undefined, visible: false })
+  const onPause = useCallback(() => setPause(!paused), [paused])
```
Comment (bafonins) [inline pkg/webui/console/components/events/index.js]:
```suggestion
  const onPause = useCallback(() => setPause(paused => !paused), [])
```

File: pkg/webui/console/components/events/list.js [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    />
+  </div>
+)
+
+EmptyMessage.propTypes = {
+  entityId: PropTypes.string.isRequired,
+}
+
+const EventsList = React.memo(
+  ({ events, paused, scoped, entityId, onRowClick, activeId }) => {
```
Comment (bafonins) [inline pkg/webui/console/components/events/list.js]: `paused` is not used

File: pkg/webui/console/components/events/previews/gateway-status.js [hunk 3]
```diff
@@ -13,31 +13,28 @@
...
-import style from './column.styl'
+import messages from '../messages'
+
+import DescriptionList from './shared/description-list'
+import JSONPayload from './shared/json-payload'
 
-const HeaderColumn = props => {
-  const { className, content } = props
+const GatewayStatusPreview = React.memo(({ event }) => {
+  const metrics = event.data.metrics
```
Comment (bafonins) [inline pkg/webui/console/components/events/previews/gateway-status.js]:
```suggestion
  const metrics = getByPath(event, 'data.metrics')
```

File: pkg/webui/console/components/events/utils/index.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+import { getDeviceId, getApplicationId } from '@ttn-lw/lib/selectors/id'
+
+import DefaultPreview from '../previews/default'
+import DefaultSyntheticEventPreview from '../previews/synthetic/default'
+import SyntheticErrorEventPreview from '../previews/synthetic/error'
+
+import { eventIconMap, dataTypeMap, applicationUpMessages } from './definitions'
+
+export const getEventId = event => `${event.time}-${event.name}`
```
Comment (bafonins) [inline pkg/webui/console/components/events/utils/index.js:23]: Just as a note: with https://github.com/TheThingsNetwork/lorawan-stack/pull/3211 we can use the `unique_id` field.

**Review by bafonins — CHANGES_REQUESTED** [review]
Looks great!

1. Horizontal scroll appears on live data with expanded side navigation. (Device)
2. For some reason I get two identical events appear (app/device/gateway):

Screenshot 2020-09-21 at 14 55 59

3. Cannot clear events when paused.
4. On screen sizes 622-670px - close button on the event sidebar adds horizontal scroll and disappears from the screen
5. Please resolve prop-types warnings in storybook

File: pkg/webui/console/components/events/utils/index.js (hunk 4 above)
Comment (kschiffer) [reply pkg/webui/console/components/events/utils/index.js:23]: Yes. I'll add as soon as this is merged into `v3.10`.

**Comment by kschiffer** [comment]
Resolved your points.

> 1.  For some reason I get two identical events appear (app/device/gateway):

Can you double-check that this one is related to this PR. Couldn't reproduce on latest `v3.9`.

**Review by bafonins — APPROVED** [review]

**Pull request merged**